share/search/index_strategy/trovesearch_denorm.py

```diff
@@ -0,0 +1,985 @@
+                'aggs': {
+                    'agg_type_iri': {'terms': {
+                        'field': f'iri_value.iri_by_propertypath.{_path_field_name((RDF.type,))}',
+                    }},
+                    'agg_value_name': {'terms': {'field': 'iri_value.value_name'}},
+                    'agg_value_title': {'terms': {'field': 'iri_value.value_title'}},
+                    'agg_value_label': {'terms': {'field': 'iri_value.value_label'}},
+                },
+            },
+        },
+    }
+
+
+def _build_date_valuesearch(params: ValuesearchParams, cursor: _SimpleCursor) -> dict:
+    assert not params.valuesearch_textsegment_set
+    assert not params.valuesearch_filter_set
+    _bool = _BoolBuilder()
+    _bool.add_boolparts(
+        _QueryHelper(
+            base_field='card',
+            textsegment_set=params.cardsearch_textsegment_set,
+            filter_set=params.cardsearch_filter_set,
+            relevance_matters=False,
+        ).boolparts(),
+    )
+    # exclude iri_value docs (possible optimization: separate indexes)
+    _bool.add_boolpart('must_not', {'exists': {'field': 'iri_value'}})
+    _field = f'card.date_by_propertypath.{_path_field_name(params.valuesearch_propertypath)}'
+    return {
+        'query': _bool.as_query(),
+        'size': 0,  # ignore hits; just want the aggs
+        'aggs': {'agg_valuesearch_dates': {
+            'date_histogram': {
+                'field': _field,
+                'calendar_interval': 'year',
+                'format': 'yyyy',
+                'order': {'_key': 'desc'},
+                'min_doc_count': 1,
+            },
+        }}
+    }
+
+
+###
+# assorted helper functions
+
+def _bucketlist(agg_result: dict) -> list[str]:
+    return [
+        _bucket['key']
+        for _bucket in agg_result['buckets']
+    ]
+
+
+def _daterange_value(datevalue: str):
+    _cleanvalue = datevalue.strip()
+    if re.fullmatch(r'\d{4,}', _cleanvalue):
+        return f'{_cleanvalue}||/y'
+    if re.fullmatch(r'\d{4,}-\d{2}', _cleanvalue):
+        return f'{_cleanvalue}||/M'
+    if re.fullmatch(r'\d{4,}-\d{2}-\d{2}', _cleanvalue):
+        return f'{_cleanvalue}||/d'
+    raise ValueError(f'bad date value "{datevalue}"')
+
+
+def _depth_field_name(depth: int) -> str:
+    return f'depth{depth}'
+
+
+def _path_field_name(path: Propertypath) -> str:
+    return ts.b64(ts.propertypath_as_keyword(path))
+
+
+def _parse_path_field_name(path_field_name: str) -> Propertypath:
+    # inverse of _path_field_name
+    _list = json.loads(ts.b64_reverse(path_field_name))
+    assert isinstance(_list, list)
+    assert all(isinstance(_item, str) for _item in _list)
+    return tuple(_list)
+
+
+def _any_query(queries: abc.Collection[dict]):
+    if len(queries) == 1:
+        (_query,) = queries
+        return _query
+    return {'bool': {'should': list(queries), 'minimum_should_match': 1}}
+
+
+###
+# cursor implementations
+
+@dataclasses.dataclass
+class _SimpleCursor:
```

Comment:
Some of the other `dataclasses` in here are very-similar-but-somewhat-different-from some of the `dataclasses` found in `indexcard_flats`, but it looks like the only change in this one as compared to `index_card_flats._SimpleCursor` is the `assert` in `from_page_param`. 

Should this (and/or others, potentially) be DRY'd up a bit?

Reply:
my intent was/is to delete `trove_indexcard_flats` as soon as `trovesearch_denorm` is in stable use, so tried to avoid putting much effort into updating it further -- but agree there are a few things (like cursors) that would make more sense reused than duplicated, yeah